### PR TITLE
Propose extract_bfs_paths C API

### PR DESCRIPTION
**Pull request by ChuckHastings**
This PR defines the C API for extract_bfs_paths.  Implementation will follow once the API is approved.

Partially addresses #1891

### CODE REVIEW

**Comment by ChuckHastings** [comment]
rerun tests

**Comment by ChuckHastings** [comment]
rerun tests

**Comment by rlratzel** [comment]
rerun tests

Rerunning to get latest updates that should fix build problem.

**Comment by rlratzel** [comment]
I talked with @ChuckHastings offline, and something we discussed is that these new APIs could be used for general path extraction - not just for extracting paths from BFS results - and should be renamed without `_bfs` in the names (and other changes to decouple from BFS, if possible). For example, a general path extraction API could be immediately useful when we add SSSP to the C API, which returns results in the same format.

File: cpp/include/cugraph_c/algorithms.h
Comment (rlratzel) [inline cpp/include/cugraph_c/algorithms.h:222]: Reminder for either a `FIXME` or an update in this PR to address [this](https://github.com/rapidsai/cugraph/pull/1955#issuecomment-974584672).

**Review by rlratzel — COMMENTED** [review]
Just added a reminder for a `FIXME` we discussed.

Also, this PR could add a unit test too, which calls a stubbed out implementation. I personally would benefit from seeing how the API is meant to be used in the context of an actual BFS call.

**Comment by rlratzel** [comment]
rerun tests

CI failed due to what looks like a problem Jenkins had in accessing github. Rerunning in case this was temporary.

**Review by rlratzel — APPROVED** [review]
Thanks, LGTM. I had some small suggestions which need not hold up my approval.

File: cpp/tests/c_api/extract_paths_test.c
Comment (ChuckHastings) [reply cpp/tests/c_api/extract_paths_test.c:54]: Fixed in latest push

File: cpp/tests/c_api/extract_paths_test.c
Comment (ChuckHastings) [reply cpp/tests/c_api/extract_paths_test.c:55]: Fixed in latest push

**Comment by BradReesWork** [comment]
rerun tests

File: cpp/include/cugraph_c/algorithms.h
Comment (seunghwak) [inline cpp/include/cugraph_c/algorithms.h:147]: Should we better use int32_t instead of int if the intention here is to really enforce 4 byte alignment?

**Review by seunghwak — APPROVED** [review]
Looks good to me, I have just a minor comment to think about.

File: cpp/include/cugraph_c/algorithms.h
Comment (ChuckHastings) [reply cpp/include/cugraph_c/algorithms.h:147]: Good point.  Will fix in the implementation PR.

**Comment by ChuckHastings** [comment]
@gpucibot merge

**Review by Ethyling — APPROVED** [review]

**Pull request merged**
